src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,405 @@
+                    KeyFactory kf = getKeyFactory(p8key.getAlgorithm());
+                    DEREncodable d;
+
+                    d = kf.generatePrivate(
+                        new PKCS8EncodedKeySpec(p8key.getEncoded(),
+                            p8key.getAlgorithm()));
+
+                    // Look for a public key inside the pkcs8 encoding.
+                    if (p8key.getPubKeyEncoded() != null) {
+                        // Check if this is a OneAsymmetricKey encoding
+                        X509EncodedKeySpec spec = new X509EncodedKeySpec(
+                            p8key.getPubKeyEncoded(), p8key.getAlgorithm());
+                        yield new KeyPair(getKeyFactory(p8key.getAlgorithm()).
+                            generatePublic(spec), (PrivateKey) d);
+
+                    } else if (d instanceof PKCS8Key p8 &&
+                        p8.getPubKeyEncoded() != null) {
+                        // If the KeyFactory decoded an algorithm-specific
+                        // encodings, look for the public key again.  This
+                        // happens with EC and SEC1-v2 encoding
+                        X509EncodedKeySpec spec = new X509EncodedKeySpec(
+                            p8.getPubKeyEncoded(), p8.getAlgorithm());
+                        yield new KeyPair(getKeyFactory(p8.getAlgorithm()).
+                            generatePublic(spec), p8);
+                    } else {
+                        // No public key, return the private key.
+                        yield d;
+                    }
+                }
+                case Pem.ENCRYPTED_PRIVATE_KEY -> {
+                    if (password == null) {
+                        yield new EncryptedPrivateKeyInfo(decoder.decode(pem.pem()));
+                    }
+                    yield new EncryptedPrivateKeyInfo(decoder.decode(pem.pem())).
+                        getKey(password.getPassword());
+                }
+                case Pem.CERTIFICATE, Pem.X509_CERTIFICATE -> {
+                    CertificateFactory cf = getCertFactory("X509");
+                    yield (X509Certificate) cf.generateCertificate(
+                        new ByteArrayInputStream(decoder.decode(pem.pem())));
+                }
+                case Pem.X509_CRL -> {
+                    CertificateFactory cf = getCertFactory("X509");
+                    yield (X509CRL) cf.generateCRL(
+                        new ByteArrayInputStream(decoder.decode(pem.pem())));
+                }
+                case Pem.RSA_PRIVATE_KEY -> {
+                    KeyFactory kf = getKeyFactory("RSA");
+                    yield kf.generatePrivate(
+                        RSAPrivateCrtKeyImpl.getKeySpec(decoder.decode(pem.pem())));
+                }
+                default -> pem;
+            };
+        } catch (GeneralSecurityException | IOException e) {
+            throw new IllegalArgumentException(e);
```

Comment:
Should all of these decoding issues really be treated as `IllegalArgumentException`s? I would think that general decoding issues should be thrown as checked exceptions. Runtime exceptions are typically used for serious issues, like programming errors.